src/sound/xaudio2-driver.cpp

```diff
@@ -0,0 +1,624 @@
+}
+
+class XAudio2SoundDriver::XAudio2GarbageCollector {
+  public:
+    XAudio2GarbageCollector(XAudio2SoundDriver& driver)
+            : _driver(driver),
+              _wake_event(nullptr),
+              _quit_event(nullptr),
+              _idle_event(nullptr),
+              _is_quitting(0),
+              _next_collectable(nullptr) {
+        _wake_event = CreateEventA(nullptr, FALSE, FALSE, nullptr);
+        _quit_event = CreateEventA(nullptr, FALSE, FALSE, nullptr);
+        _idle_event = CreateEventA(nullptr, FALSE, FALSE, nullptr);
+
+        if (_wake_event == nullptr || _quit_event == nullptr || _quit_event == nullptr)
+            throw std::runtime_error("CreateEvent failed");
+
+        if (!CreateThread(nullptr, 0, XAudio2GarbageCollector::thread_start, this, 0, nullptr))
+            throw std::runtime_error("CreateThread failed");
+
+        _is_started = true;
+    }
+
+    ~XAudio2GarbageCollector() {
+        if (_is_started)
+            this->shutdown();
+
+        if (_quit_event)
+            CloseHandle(_quit_event);
+
+        if (_wake_event)
+            CloseHandle(_wake_event);
+
+        if (_idle_event)
+            CloseHandle(_idle_event);
+    }
+
+    void link_object(XAudio2GarbageCollectable* obj) {
+        // Put this at the head of the list
+        for (;;) {
+            XAudio2GarbageCollectable* next = _next_collectable;
+            obj->set_next_gc(next);
+            XAudio2GarbageCollectable* cas_result = cas_collectable(_next_collectable, next, obj);
+            if (cas_result == next)
+                break;
+        }
+    }
+
+  private:
+    void shutdown() {
+        // Wait for any partial GC cycle to finish
+        ResetEvent(_idle_event);
+        SetEvent(_wake_event);
+        WaitForSingleObject(_idle_event, INFINITE);
+
+        // Run one more cycle
+        SetEvent(_wake_event);
+        WaitForSingleObject(_idle_event, INFINITE);
+
+        // Quit
+        _is_quitting = true;
+        SetEvent(_wake_event);
+
+        // Wait for the thread to quit
+        WaitForSingleObject(_quit_event, INFINITE);
+    }
+
+    XAudio2GarbageCollectable* atomic_swap_chain_head(XAudio2GarbageCollectable* new_head) {
+        return static_cast<XAudio2GarbageCollectable*>(InterlockedExchangePointer(
+                reinterpret_cast<PVOID volatile*>(&_next_collectable), new_head));
+    }
+
+    static XAudio2GarbageCollectable* cas_collectable(
+        XAudio2GarbageCollectable*volatile& dest, XAudio2GarbageCollectable* expected,
+        XAudio2GarbageCollectable* replacement) {
+        return static_cast<XAudio2GarbageCollectable*>(InterlockedCompareExchangePointer(
+                reinterpret_cast<PVOID volatile*>(&dest), replacement, expected));
+    }
+
+    int thread_func() {
+        for (;;) {
+            WaitForSingleObject(_wake_event, INFINITE);
+
+            XAudio2GarbageCollectable* next = atomic_swap_chain_head(nullptr);

```

Comment:
I guess it might not even be necessary to do all the GC stuff... The point was a.) to not do memory management on the audio thread, but that only matters if the sound is unloaded while still being played, and b.) to not block the thread when calling DestroyVoice, but that also only matters if a channel is deleted while something else is happening, and I guess neither is terribly common in this case.